neuraxle/distributed/streaming.py

```diff
@@ -0,0 +1,510 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+"""
+from abc import abstractmethod
+from multiprocessing import Queue, Lock
+from multiprocessing.context import Process
+from threading import Thread
+from typing import Tuple, List, Union, Iterable
+
+from neuraxle.base import NamedTupleList, ExecutionContext, BaseStep, MetaStepMixin, NonFittableMixin, FullDumpLoader
+from neuraxle.data_container import DataContainer, ListDataContainer
+from neuraxle.pipeline import Pipeline, CustomPipelineMixin
+from neuraxle.steps.numpy import NumpyConcatenateOuterBatch
+
+
+class ObservableQueue:
+    def __init__(self, queue):
+        self.queue = queue
+        self.observers = []
+
+    def subscribe(self, queue_worker: 'ObservableQueue') -> 'ObservableQueue':
+        self.observers.append(queue_worker.queue)
+        return self
+
+    def put(self, value):
+        self.queue.put(value)
+
+    def notify(self, value):
+        for observer in self.observers:
+            observer.put(value)
+
+
+class QueuedPipelineTask(object):
+    def __init__(self, step_name, data_container):
+        self.step_name = step_name
+        self.data_container = data_container
+
+
+class QueueWorker(ObservableQueue, MetaStepMixin, BaseStep):
+    """
+    Start multiple Process or Thread that process items from the queue of batches to process.
+    It is both an observable, and observer.
+    It notifies the results of the wrapped step handle transform method.
+    It receives the next data container to process.
+
+    .. seealso::
+        :class:`Observer`,
+        :class:`Observable`,
+        :class:`MetaStepMixin`,
+        :class:`BaseStep`
+    """
+
+    def __init__(
+            self,
+            wrapped: BaseStep,
+            max_size: int,
+            n_workers: int,
+            use_threading: bool,
+            additional_worker_arguments=None,
+            use_savers=False
+    ):
+        if additional_worker_arguments is None:
+            additional_worker_arguments = []
+
+        MetaStepMixin.__init__(self, wrapped)
+        BaseStep.__init__(self)
+        ObservableQueue.__init__(self, Queue(maxsize=max_size))
+
+        self.use_threading: bool = use_threading
+        self.workers: List[Process] = []
+        self.n_workers: int = n_workers
+        self.observers: List[Queue] = []
+        self.additional_worker_arguments = additional_worker_arguments
+        self.use_savers = use_savers
+
+    def start(self, context: ExecutionContext):
+        """
+        Start multiple processes or threads with the worker function as a target.
+
+        :param context: execution context
+        :type context: ExecutionContext
+        :return:
+        """
+        target_function = worker_function
+        if self.use_savers:
+            self.wrapped.save(context, full_dump=True)
```

Comment:
I need the full dump savers pull request merged for this ^^^

Reply:
I will remove the use_savers for now. ill put back those lines when the pr is merged.